src/middlewared/middlewared/plugins/apps/crud.py

```diff
@@ -248,21 +249,32 @@ def create_internal(
                 compose_action(app_name, version, 'up', force_recreate=True, remove_orphans=True)
         except Exception as e:
             job.set_progress(80, f'Failure occurred while installing {app_name!r}, cleaning up')
-            for method, args, kwargs in (
-                (compose_action, (app_name, version, 'down'), {'remove_orphans': True}),
-                (shutil.rmtree, (get_installed_app_path(app_name),), {}),
-            ):
-                with contextlib.suppress(Exception):
-                    method(*args, **kwargs)
-
-            self.middleware.call_sync('app.metadata.generate').wait_sync(raise_error=True)
+            # We only want to remove app volume ds if it did not exist before the installation
+            # and was created during this installation process
+            self.remove_failed_resources(app_name, version, app_volume_ds_exists is False)
             self.middleware.send_event('app.query', 'REMOVED', id=app_name)
             raise e from None
         else:
             if dry_run is False:
                 job.set_progress(100, f'{app_name!r} installed successfully')
                 return self.get_instance__sync(app_name)
 
+    @private
+    def remove_failed_resources(self, app_name, version, remove_ds=False):
+        apps_volume_ds = self.get_app_volume_ds(app_name) if remove_ds else None
+
+        with contextlib.suppress(Exception):
+            compose_action(app_name, version, 'down', remove_orphans=True)
+
+        shutil.rmtree(get_installed_app_path(app_name), ignore_errors=True)
+
+        if apps_volume_ds and remove_ds:
+            with contextlib.suppress(Exception):
```

Comment:
Maybe at least log this so we know why it failed if it did as the dataset should exist based on above func call